accounts-db/src/file_io.rs

```diff
@@ -68,3 +68,77 @@ pub fn read_buffer(
 ) -> std::io::Result<usize> {
     panic!("unimplemented");
 }
+
+#[cfg(test)]
+mod tests {
+
+    use {super::*, std::io::Write, tempfile::tempfile};
+
+    #[cfg(unix)]
+    #[test]
+    fn test_read_buffer() {
+        // Setup a sample file with 32 bytes of data
+        let mut sample_file = tempfile().unwrap();
+        let bytes: Vec<u8> = (0..32).collect();
+        sample_file.write_all(&bytes).unwrap();
+
+        // Read all 32 bytes into buffer
+        let mut buffer = [0; 32];
+        let num_bytes_read = read_buffer(&sample_file, 0, &mut buffer, 32).unwrap();
+        assert_eq!(num_bytes_read, 32);
+        assert_eq!(bytes, buffer);
+
+        // Given a 64-byte buffer, it should only read 32 bytes into the buffer
+        let mut buffer = [0; 64];
+        let num_bytes_read = read_buffer(&sample_file, 0, &mut buffer, 32).unwrap();
+        assert_eq!(num_bytes_read, 32);
+        assert_eq!(bytes, buffer[0..32]);
+        assert_eq!(buffer[32..64], [0; 32]);
+
+        // Given the `valid_len` is 16, it should only read 16 bytes into the buffer
+        let mut buffer = [0; 32];
+        let num_bytes_read = read_buffer(&sample_file, 0, &mut buffer, 16).unwrap();
+        assert_eq!(num_bytes_read, 16);
+        assert_eq!(bytes[0..16], buffer[0..16]);
+        // As a side effect of the `read_buffer` the data passed `valid_len` was
+        // read and put into the buffer, though these data should not be
+        // consumed.
+        assert_eq!(buffer[16..32], bytes[16..32]);
+
+        // Given the start offset 8, it should only read 24 bytes into buffer
+        let mut buffer = [0; 32];
+        let num_bytes_read = read_buffer(&sample_file, 8, &mut buffer, 32).unwrap();
+        assert_eq!(num_bytes_read, 24);
+        assert_eq!(buffer[0..24], bytes[8..32]);
+        assert_eq!(buffer[24..32], [0; 8])
+    }
+
+    #[cfg(unix)]
+    #[test]
+    fn test_read_more_buffer() {
```

Comment:
i don't know what conditions would cause a read to return partial results. Maybe if file is big? maybe big is multiple pages? I don't know. I know the internal read api says it can return partial results. Just trying to see if we can test the loop inside read/read_more_buffer that keeps working to fill up the buffer. any ideas?

Reply:
EINTR and EWOULDBLOCK. We can ignore the latter since we never set the files as non blocking, but we must handle `ErrorKind::Interrupted`

Reply:
@alessandrod here is my proposed change:
```
pub fn read_buffer(
    file: &File,
    start_offset: usize,
    buffer: &mut [u8],
    valid_len: usize,
) -> std::io::Result<usize> {
    let mut offset = start_offset;
    let mut start_read = 0;
    let mut total_bytes_read = 0;
    if start_offset >= valid_len {
        return Ok(0);
    }

    while start_read < buffer.len() {
        match file.read_at(&mut buffer[start_read..], offset as u64) {
            Err(err) => {
                if err.kind() == std::io::ErrorKind::Interrupted {
                    continue;
                }
                return Err(err);
            }
            Ok(bytes_read_this_time) => {
                total_bytes_read += bytes_read_this_time;
                if total_bytes_read + start_offset >= valid_len {
                    total_bytes_read -= (total_bytes_read + start_offset) - valid_len;
                    // we've read all there is in the file
                    break;
                }
                // There is possibly more to read. `read_at` may have returned partial results, so prepare to loop and read again.
                start_read += bytes_read_this_time;
                offset += bytes_read_this_time;
            }
        }
    }
    Ok(total_bytes_read)
}
```

Reply:
I went ahead and submitted this. @alessandrod would love for you to critique it thoroghly.

Reply:
left some comments!